File: client/src/game/systems/inputmovement.lua
local Aspect = require 'src.ecs.aspect'
local System = require 'src.ecs.system'
local Fixture = require 'src.game.components.fixture'
local Input = require 'src.game.components.input'
local Movement = require 'src.game.components.movement'
local Position = require 'src.game.components.position'
local Respawn = require 'src.game.components.respawn'

local aspect = Aspect.new({Fixture, Input, Movement, Position}, {Respawn})
local InputMovement = System:new('inputmovement', aspect)

function InputMovement:update(dt)
  for _, entity in pairs(self.entities) do
    local fixture = entity:as(Fixture)
    local movement = entity:as(Movement)
    local body = fixture.fixture:getBody()
    local velocityX, velocityY = body:getLinearVelocity()

    local newVelocityX = 0
    local newVelocityY = velocityY

    if (movement.left == true) then
      newVelocityX = -300
    elseif movement.right == true then
      newVelocityX = 300
    end

    if (movement.jump and movement.jumpCount < 2) then
      newVelocityY = -1000
      movement.jump = false
      movement.jumpCount = movement.jumpCount + 1
    end

    body:setLinearVelocity(newVelocityX, newVelocityY)
  end
end

return InputMovement

File: client2/src/game/systems/InputMovement.ts
import { Aspect } from '../../ecs/Aspect'
import { System } from '../../ecs/System'
import { GameObject } from '../components/GameObject'
import { Input } from '../components/Input'
import { Movement } from '../components/Movement'
import { Position } from '../components/Position'
import { Respawn } from '../components/Respawn'
import { SystemFlag } from '../flags'

export class InputMovementSystem extends System {
  static _id = 'InputMovement'
  _id = InputMovementSystem._id

  static _flag = SystemFlag.InputMovement
  _flag = InputMovementSystem._flag

  static _aspect = new Aspect(
    [GameObject, Input, Movement, Position],
    [Respawn]
  )
  _aspect = InputMovementSystem._aspect

  update = (dt: number) => {
    this.entities.forEach(entity => {
      const gameObject = entity.as(GameObject)
      const movement = entity.as(Movement)

      if (!gameObject || !movement) {
        return
      }

      const body = gameObject.fixture.getBody()
      let [velocityX, velocityY] = body.getLinearVelocity()

      if (movement.left) {
        velocityX = -300
      } else if (movement.right) {
        velocityX = 300
      }

      if (movement.jump && movement.jumpCount < 2) {
        velocityY = -1000
        movement.jump = false
        movement.jumpCount++
      }

      body.setLinearVelocity(velocityX, velocityY)
    })
  }
}
